### DISCUSSION TITLE
How to export a multi layer TIF file, keep the original layer name to generate PNG, and not crop the layers

### DISCUSSION THREAD

**Discussion by 291598912** [ImageMagick/ImageMagick#6301] [Digital Image Processing]
### ImageMagick version

7.1.0.10

### Operating system, version and so on

windows11

### Description

多图层的tif文件，怎么导出图层，保留原来图层名字生成png，且不裁剪图层

**Comment by 291598912** [comment]
How to export a multi layer TIF file, keep the original layer name to generate PNG, and not crop the layers

**Comment by snibgo** [comment]
Suppose file `in.tiff` contains images with labels. Then this command will create one PNG file per input image. Each PNG file will be named for the label, with `.png` extension.

```
magick in.tiff -set filename:MYFILE %l %[filename:MYFILE].png
```
If using in a Windows BAT script, double each `%`.

Does that do what you want?

**Reply by 291598912** [reply]
This Chinese layer name may have garbled characters，But I still cropped the layer. I want to keep the layer in its original background position

**Reply by 291598912** [reply]
Can the png obtained in this way be the same as the original input.tif in synthesis

**Comment by 291598912** [comment]
I would like to consult some teachers about the CMYK color mode multi layer TIF file. Extracting layers will result in color changes. How can this be resolved? I can't even try to convert it to RGB.
Im_.png is the layer obtained by magick splitting tif
Ps_.png is the layer split by Photoshop
![im_](https://user-images.githubusercontent.com/16145102/236738040-7e9d7092-5289-45a6-a5f6-e9f7214a1a49.png)
![ps_](https://user-images.githubusercontent.com/16145102/236738046-58d60877-4a96-42ad-85d1-99456a030edd.png)

The command line is:
magick input.tif %d.png

**Reply by 291598912** [reply]
The original tif file is:
[intput.zip](https://github.com/ImageMagick/ImageMagick/files/11417907/intput.zip)

**Comment by snibgo** [comment]
The file `intput.tiff` contains 23 CMYK images, each with an embedded ICC colour profile. PNG cannot record CMYK profiles. If you want to convert them to sRGB, use "`-profile sRGB.icc`" or the filename of whatever sRGB profile you want. If the ICC file is not in your current directory, provide the full path.

**Reply by 291598912** [reply]
The issue of color difference has been resolved, - profile sRGB. icc, but there are still issues with cropping

**Comment by snibgo** [comment]
You can see what sizes IM thinks the images are:
```
magick intput.tif info:

intput.tif[0] TIFF 1122x2209 1122x2209+0+0 8-bit ColorSeparation CMYK 12.4279MiB 0.016u 0:00.009
intput.tif[1] TIFF 924x1874 924x1874+10+36 8-bit CMYK 12.4279MiB 0.016u 0:00.012
intput.tif[2] TIFF 924x1874 924x1874+10+36 8-bit CMYK 12.4279MiB 0.016u 0:00.013
intput.tif[3] TIFF 916x1866 916x1866+18+44 8-bit CMYK 12.4279MiB 0.016u 0:00.013
intput.tif[4] TIFF 916x1866 916x1866+18+44 8-bit CMYK 12.4279MiB 0.016u 0:00.014
intput.tif[5] TIFF 917x1867 917x1867+17+43 8-bit CMYK 12.4279MiB 0.016u 0:00.015
intput.tif[6] TIFF 888x1800 888x1800+12+76 8-bit CMYK 12.4279MiB 0.016u 0:00.016
intput.tif[7] TIFF 888x1800 888x1800+12+76 8-bit CMYK 12.4279MiB 0.016u 0:00.016
intput.tif[8] TIFF 524x259 524x259+10+36 8-bit CMYK 12.4279MiB 0.016u 0:00.017
intput.tif[9] TIFF 226x31 226x31+161+41 8-bit CMYK 12.4279MiB 0.016u 0:00.018
intput.tif[10] TIFF 226x31 226x31+554+51 8-bit CMYK 12.4279MiB 0.016u 0:00.019
intput.tif[11] TIFF 226x31 226x31+144+1873 8-bit CMYK 12.4279MiB 0.016u 0:00.020
intput.tif[12] TIFF 31x226 31x226+24+307 8-bit CMYK 12.4279MiB 0.016u 0:00.020
intput.tif[13] TIFF 31x226 31x226+24+703 8-bit CMYK 12.4279MiB 0.016u 0:00.021
intput.tif[14] TIFF 31x226 31x226+24+1094 8-bit CMYK 12.4279MiB 0.016u 0:00.022
intput.tif[15] TIFF 31x226 31x226+24+1470 8-bit CMYK 12.4279MiB 0.016u 0:00.023
intput.tif[16] TIFF 31x226 31x226+897+148 8-bit CMYK 12.4279MiB 0.000u 0:00.023
intput.tif[17] TIFF 31x226 31x226+897+405 8-bit CMYK 12.4279MiB 0.000u 0:00.025
intput.tif[18] TIFF 31x226 31x226+897+801 8-bit CMYK 12.4279MiB 0.000u 0:00.027
intput.tif[19] TIFF 31x226 31x226+897+1192 8-bit CMYK 12.4279MiB 0.000u 0:00.028
intput.tif[20] TIFF 31x226 31x226+897+1568 8-bit CMYK 12.4279MiB 0.000u 0:00.028
intput.tif[21] TIFF 226x31 226x31+570+1873 8-bit CMYK 12.4279MiB 0.000u 0:00.029
intput.tif[22] TIFF 226x31 226x31+162+259 8-bit CMYK 12.4279MiB 0.000u 0:00.030
```
For example, IM thinks the first image is 1122x2209 pixels. Gimp agrees.

Which of these are wrong?

**Reply by 291598912** [reply]
What I mean is that starting from the first layer, the image size is 1122x2209

**Reply by 291598912** [reply]
The exported layers are all 1122x2209 sized

**Reply by 291598912** [reply]
![8](https://user-images.githubusercontent.com/16145102/236856748-e0217c96-e05e-4467-a086-329a7eab0c7a.png)
![镜头框](https://user-images.githubusercontent.com/16145102/236856775-2c42756d-030f-4977-86b6-79deec2818a3.png)
What I want is the size of the second image

**Reply by 291598912** [reply]
The picture posted above：
intput.tif[8] TIFF 524x259 524x259+10+36 8-bit CMYK 12.4279MiB 0.016u 0:00.017

**Reply by 291598912** [reply]
convert intput.tif -profile sRGB.icc -depth 8 -channel rgba %d.png
convert intput.tif -profile sRGB.icc -depth 16  %d.png
convert intput.tif -profile sRGB.icc -depth 32  %d.png
Regardless of the setting, it is not 32-bit, either 2-bit, 8-bit, or 48,64-bit

**Comment by snibgo** [comment]
> The exported layers are all 1122x2209 sized

That is not what I get.

If we do a simple conversion:
```
magick intput.tif x.png
```
... then we get files like this:

x-0.png 1122x2209 pixels
x-1.png 924x1874 pixels

...and so on.

**Reply by 291598912** [reply]
The meaning can only be extracted from the actual size of the layer and the generated image

**Reply by 291598912** [reply]
Then I will obtain the coordinate positions of the layers and generate them myself

**Comment by fmw42** [comment]
You can crop using the virtual canvas coordinates, e.g. for the last image which is 
intput.tif[22] TIFF 226x31 226x31+162+259 8-bit CMYK 12.4279MiB 0.000u 0:00.030
use

`magick intput.tif[22] -crop 226x31+162+259 x.png`

Does that do what you want?

**Reply by 291598912** [reply]
This is not what I want, it doesn't need to be cropped

**Comment by 291598912** [comment]
Can the png obtained in this way be the same as the original input.tif in synthesis

**Reply by CrewS** [reply]
Have you done it, composited the png back into the same Tif?

**Comment by 291598912** [comment]
Why do many PNGs have depths of 2 or 8 bits, can they be 32 bits

**Comment by snibgo** [comment]
> Can the png obtained in this way be the same as the original input.tif in synthesis

I don't understand the question. Each output PNG should be the same width and height, in pixels, as the corresponding image from the input TIF file.

> Why do many PNGs have depths of 2 or 8 bits, can they be 32 bits

PNG files can have 1, 2, 4, 8, or 16 bits per channel.

**Comment by 291598912** [comment]
The exported RGB image still has color difference
convert intput.tif -profile sRGB.icc %d.png

**Reply by 291598912** [reply]
The color difference has been resolved, but I still don't know how to use commands to generate 32-bit images

**Comment by snibgo** [comment]
What do you mean by "32-bit images"? PNG files cannot have 32 bits per channel. But they can have 8 bits per channel, with four channels (RGBA), making a total of 32 bits per pixel.

**Reply by 291598912** [reply]
Bit depth is always 8 or 2 bits

**Reply by 291598912** [reply]
It should be using png32 to write files. How can I handle this

**Comment by fmw42** [comment]
Perhaps you want

`convert intput.tif -profile sRGB.icc PNG32:%d.png`

the PNG32: preface forces a 32-bit (4*8-bit channels) output

PNG does not support 32-bits/channel only 32-bits per image with each of 4 channels equal to 8-bits
